File: 1122-relative-sort-array/main.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        
        order_map = {num: idx for idx, num in enumerate(arr2)}

        count_map = {}
        for num in arr1:
            count_map[num] = count_map.get(num, 0) + 1


        ans = [] 
        for num in arr2:
            if num in count_map:
                ans.extend([num] * count_map[num])
                del count_map[num]
        

        remaining = sorted([num for num in count_map for _ in range(count_map[num])])
        ans.extend(remaining)


        return ans

File: 1122-relative-sort-array/main.rs
use std::collections::HashMap;
use std::iter::repeat;

impl Solution {
    pub fn relative_sort_array(arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
        
        let mut order_map: HashMap<i32, usize> = HashMap::new();
        for (index, &value) in arr2.iter().enumerate() {
            order_map.insert(value, index);   
        }

        let mut count_map: HashMap<i32, i32> = HashMap::new();
        for &num in arr1.iter() {
            *count_map.entry(num).or_insert(0) += 1;
        }

        let mut ans: Vec<i32> = Vec::new();
        for &num in arr2.iter() {
            if let Some(&count) = count_map.get(&num) {
                ans.extend(repeat(num).take(count as usize));
                count_map.remove(&num);
            }
        }

        let mut remaining: Vec<i32> = Vec::new();
        for (&num, &count) in count_map.iter() {
           remaining.extend(repeat(num).take(count as usize));
        }
        remaining.sort();

        ans.extend(remaining);

        ans
    }
}
